### PR TITLE
simplification and improvement of disk_buffer_pool

### PR DESCRIPTION
**Pull request by arvidn**

### CODE REVIEW

File: src/disk_buffer_pool.cpp [hunk 1]
```diff
@@ -364,148 +288,70 @@ namespace libtorrent
...
-			m_using_pool_allocator = m_want_pool_allocator;
-#endif
-
-#if TORRENT_HAVE_MMAP
-		// if we've already allocated an mmap, we can't change
-		// anything unless there are no allocations in use
+		// we can only change the cache size when it's unused. Since the whole
+		// cache is allocated as a single blob, to support resizing otherwise
+		// would require creating copies of every block (at least dirty blocks)
+		// and update all pointers in the block_cache.
```
Comment (ssiloti) [inline src/disk_buffer_pool.cpp]: In light of this, you may want to adjust the size limit for 32-bit systems. In my experience, attempting to allocate more than ~1GB of contiguous space on 32-bit systems usually fails due to address space fragmentation.

**Comment by arvidn** [comment]
The main challenges with this patch are:

1. making bt_peer_connection and web_peer_connection support disk buffer allocation to fail, and try again later.

2. what to do when the cache size is set to 0. We no longer can guarantee that each peer gets at least one disk block each, because the total size is fixed. The naive solution would be to set a lower limit to whatever the normal number of peers is.

File: src/bitfield.cpp [hunk 2]
```diff
@@ -190,5 +190,118 @@ namespace libtorrent
...
+		if (words == 0) return -1;
+		std::uint32_t const* const ptr = &m_buf[1];
+		for (int i = 0; i < words; ++i)
+		{
+			if (ptr[i] == 0) continue;
+
+			// now, find the index of the first set bit in ptr[i]
+			uint32_t v = aux::network_to_host(ptr[i]);
+
+#if TORRENT_HAS_SSE
```
Comment (aldenml) [inline src/bitfield.cpp]: just to leave a note, now there is `TORRENT_HAS_BUILTIN_CLZ` and can be seen in use at `sha1_hash::count_leading_zeroes`

File: src/sha1_hash.cpp [hunk 3]
```diff
@@ -95,7 +94,7 @@ namespace libtorrent
 			v |= v >> 8;
 			v |= v >> 16;
 
-			return ret + MultiplyDeBruijnBitPosition[
+			return ret + 31 - MultiplyDeBruijnBitPosition[
```
Comment (aldenml) [inline src/sha1_hash.cpp]: hmm...do you think this is a bug that could be affecting arm/android binaries using RC_1_1?

File: src/bitfield.cpp [hunk 4]
```diff
@@ -190,5 +190,124 @@ namespace libtorrent
...
+
+		// turn it into a problem of finding the last _set_ bit instead
+		last ^= mask;
+
+		// if no bit is _set_ in this word, keep scanning full words in the
+		// loop below
+		if (last != 0)
+		{
+			int const i = words - 1;
+#if TORRENT_HAS_SSE
```
Comment (aldenml) [inline src/bitfield.cpp]: just in case, another note about `TORRENT_HAS_BUILTIN_CLZ` and a few lines below

File: src/sha1_hash.cpp (hunk 3 above)
Comment (arvidn) [inline src/sha1_hash.cpp]: I checked RC_1_1, but this function doesn't exist there

File: src/bitfield.cpp (hunk 4 above)
Comment (arvidn) [inline src/bitfield.cpp]: The original intent was for the first call to __buildin_clz to really emit the assembly instruction. I realize what that intrinsic does is actually generate either the SSE instruction or the work-around computation. Presumably based on the -march argument. The intention is to make the decision at run-time whether SSE is supported and then execute the SSE instruction, or fall back to a regular computation

Comment (aldenml) [inline src/bitfield.cpp]: well, It is my understanding that `clz` (and friends) are not SSE instructions, and additional testing for `mmx_support` will prevent any ARM optimization

File: src/sha1_hash.cpp (hunk 3 above)
Comment (aldenml) [inline src/sha1_hash.cpp]: thanks, good to know :)

File: src/bitfield.cpp (hunk 4 above)
Comment (arvidn) [inline src/bitfield.cpp]: presumably there's some way to detect at runtime whether they are supported by the CPU or not. I was under the impression that checking SSE version is how one would do this. I don't see how SSE specific code could prevent ARM optimizations. ARM specific code should go side-by-side with SSE code presumably.

Comment (aldenml) [inline src/bitfield.cpp]: I rather not get you more noise with this, this PR is already delicate enough :)

I can take a look at the assembly part, specially the `ctz` that is not readily available in ARM, once you merge this

**Comment by aldenml** [comment]
with the amount of conflicts in this PR, what do you think about moving pieces until it's easier to create a new PR?

**Comment by arvidn** [comment]
hat do you mean by "moving pieces"? I intend to rebase this on top of master

**Comment by aldenml** [comment]
the little pieces like the `tailqueue` changes, I tested the rebase and it failed miserable, also tested merge squash to easy the conflicts but it was very problematic. Well, I assume that depends on the "git skills" :)

**Comment by arvidn** [comment]
yeah, I suppose I could cherry-pick some parts of this out into separate PRs.

File: test/test_tailqueue.cpp [hunk 5]
```diff
@@ -51,6 +51,10 @@ void check_chain(tailqueue<test_node>& chain, char const* expected)
 		i.next();
 		++expected;
 	}
+	if (!chain.empty())
+	{
+		TEST_CHECK(chain.last() == NULL || chain.last()->next == NULL);
```
Comment (aldenml) [inline test/test_tailqueue.cpp]: `nullptr`?

File: src/disk_buffer_pool.cpp [hunk 6]
```diff
@@ -352,166 +303,77 @@ namespace libtorrent
...
-					if (phys_ram > 1 * gb)
-					{
-						result += (phys_ram - 1 * gb) / 20;
-						phys_ram = 1 * gb;
-					}
-					result += phys_ram / 10;
-					m_max_use = result / m_block_size;
-				}
+			std::uint64_t const phys_ram = total_physical_ram();
+			std::int64_t const gb = 1024 * 1024 * 1024;
```
Comment (aldenml) [inline src/disk_buffer_pool.cpp]: as a note, I suppose this could be `constexpr`

File: src/disk_buffer_pool.cpp (hunk 1 above)
Comment (aldenml) [inline src/disk_buffer_pool.cpp]: hi @arvidn, this comment above, what do you think about it? I would say that the majority of Android devices are 32-bit

Comment (arvidn) [inline src/disk_buffer_pool.cpp]: I followed steven's advice, see line 338. Perhaps I should lower the limit even more though

Comment (aldenml) [inline src/disk_buffer_pool.cpp]: thanks

**Pull request closed without merging**